pkg/ast/spl/spl.peg

```diff
@@ -40,6 +40,130 @@ func ConvertCustomDateTimeFormatToEpochMs(dateStr string) (int64, error) {
     return t.UnixMilli(), nil
 }
 
+
+func applySnap(snap string, t time.Time) (time.Time, error) {
+    sec := t.Second()
+    min := t.Minute()
+    hour := t.Hour()
+    day := t.Day()
+    week := t.Weekday()
+    mon := t.Month()
+    year := t.Year()
+    
+    if snap[0] != 'w' {
+        tunit, err := strconv.Atoi(snap)
+        if err != nil {
+            return t, fmt.Errorf("Error while converting the snap: %v to integer, err: %v", snap, err)
+        }
+
+        switch utils.TimeUnit(tunit) {   
+            case utils.TMSecond:
+                return time.Date(year, mon, day, hour, min, sec, 0, time.Local), nil
+            case utils.TMMinute:
+                return time.Date(year, mon, day, hour, min, 0, 0, time.Local), nil
+            case utils.TMHour:
+                return time.Date(year, mon, day, hour, 0, 0, 0, time.Local), nil
+            case utils.TMDay:
+                return time.Date(year, mon, day, 0, 0, 0, 0, time.Local), nil
+            case utils.TMWeek:
+                diff := week - time.Sunday
+                return time.Date(year, mon, day-int(diff), 0, 0, 0, 0, time.Local), nil
+            case utils.TMMonth:
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMQuarter:
+                if mon >= time.October {
+                    mon = time.October
+                } else if mon >= time.July {
+                    mon = time.July
+                } else if mon >= time.April {
+                    mon = time.April
+                } else {
+                    mon = time.January
+                }
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMYear:
+                return time.Date(year, 1, 1, 0, 0, 0, 0, time.Local), nil
+            default:
+                return t, fmt.Errorf("Unsupported time unit for relative timestamp: %v", tunit)
+        }
+    } else {
+        if len(snap) != 2 {
+            return t, fmt.Errorf("Error for special week snap, should be of type w[0-7] got: %v", snap)
```

Comment:
nit: I think `should be of type w[0-7]` is a bit unclear; I think what you mean is `should follow the regex "w[0-7]"`, right?

Reply:
yes basically they could be from w0, w1, ... to w7